Nand.py

```diff
@@ -158,3 +192,19 @@ def delay(self):
 
 def pprint_state(state):
     return '{' + ', '.join(f"{c}.{n}: {v}" for (c, n), v in state.items()) + '}'
+
+def lookup_state_bit(state, ref):
```

Comment:
I think this is a cool trick and fits the subject matter well, but it could probably be embedded in the InputRef or split between a class for the state and the InputRef somehow.

Perhaps I'm conceptualizing the transistors wrong but I think of a multi-bit input as a collection of bits rather than N individual bits, which is how it seems to be represented here.

As a general comment on code style, it's a bit hard to follow how things are being used/invoked due to all the hidden function calls via `__function__` style, so while I appreciate the usability from a caller perspective it's taking me a bit to review.

Reply:
Yes, this is getting truly ugly at this point. I'm still planning to refactor some of it into something like `State` when I need to be able to do stateful evaluation (project 03).

I'm willing to accept the _implementation_ being hard to follow if it leads to a nice clean syntax for the solutions, which is working fairly well at the moment. On the other hand, a trivial error in a solution (like leaving off `.out`) leads to an incomprehensible runtime error. We can probably improve that by adding even more `__foo__` stuff just raise errors with hints like "did you forget ...". I'm going to put that off until the implementation is complete and we start looking for users to tell us if this is actually any good.